### PR TITLE
feat: add allowance spend_vault module

### LINKED ISSUE #178: [Feature]: Implement Allowance/Approval
Implement the design from #177.

### PR DESCRIPTION
**Pull request by kosedogus**
Resolves #178

Validity checks performed:
- `/sui-btt` in ultracode
- `/sui-basic-review` in ultracode

Coverage note: the funds-accumulator paths (`withdraw_all`, the
`&AccumulatorRoot`-taking reads, and the native pool-short abort) cannot run in
the Move unit-test VM (it cannot construct an `AccumulatorRoot`); they are
exercised by a separate on-chain test harness, available on request.

#### PR Checklist

- [x] Tests
- [x] Documentation
- [x] Changelog

## Summary by CodeRabbit

* **New Features**
  * Introduced the OpenZeppelin Allowance package, featuring a capability-keyed, pooled multi-coin `spend_vault` with budget limits, optional expiration, and revocation/suspend controls.
  * Added keeper and direct-delegation example flows for integrating the vault.
* **Documentation**
  * Updated the changelog and contracts README with the new `openzeppelin_allowance` package and detailed usage/security guidance.
* **Tests**
  * Added extensive unit tests covering vault lifecycle, funding/spending, reads, revocation, isolation/composability, and withdraw behavior.
* **Chores**
  * Expanded CI lint/test matrix to include the new Allowance package.

### CODE REVIEW

File: contracts/allowance/examples/spend_vault/defi_keeper.move [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+// === View helpers ===
+
+/// The vault this service is pinned to.
+public fun vault_id(s: &Service): ID {
+    s.vault_id
+}
+
+/// Whether `user` currently has a cap in custody.
+public fun is_registered(s: &Service, user: address): bool {
+    s.caps.contains(user)
```
[the remark this answers is not in the document]
Comment (kosedogus) [reply contracts/allowance/examples/spend_vault/defi_keeper.move]: This is an example code to show how the library can be integrated. Hence, I am not following the STYLEGUIDE 1-to-1 in order to not bloat the code.

File: contracts/allowance/examples/spend_vault/direct_delegation.move [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+    vault.withdraw_all<T>(owner_cap, root, ctx)
+}
+
+/// Tear down the vault. PRECONDITION: every coin already drained via
+/// `withdraw_all<T>` (enumerate types off-chain with `suix_getAllBalances`), or any
+/// remaining funds strand permanently at the dead vault address. `destroy` drains
+/// only the budget ledger and deletes the UIDs; it does not touch the pool.
+public fun tear_down(vault: Vault, owner_cap: OwnerCap, ctx: &mut TxContext) {
+    vault.destroy(owner_cap, ctx);
+}
```
[the remark this answers is not in the document]
Comment (kosedogus) [reply contracts/allowance/examples/spend_vault/direct_delegation.move]: This is an example code to show how the library can be integrated. Hence, I am not following the STYLEGUIDE 1-to-1 in order to not bloat the code.

File: contracts/allowance/examples/spend_vault/example_coin.move [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    transfer::public_transfer(coins, ctx.sender());
+}
+
+// === Test-Only Helpers ===
+
+/// Run `init` under test, minting the fixed supply to the sender so an example vault
+/// can be funded.
+#[test_only]
+public fun init_for_testing(ctx: &mut TxContext) {
+    init(EXAMPLE_COIN {}, ctx)
```
[the remark this answers is not in the document]
Comment (kosedogus) [reply contracts/allowance/examples/spend_vault/example_coin.move:49]: This is an example code to show how the library can be integrated. Hence, I am not following the STYLEGUIDE 1-to-1 in order to not bloat the code.

**Comment by bidzyyys** [comment]
### Review summary (`/sui-basic-review`, core `spend_vault.move`)

Exceptionally well-engineered and documented — no Critical/High/Medium correctness or security bug found; all 17 inferred invariants hold. Items worth attention:

- [x] **`destroy` permanently strands an undrained pool** (Medium, by-design) — `destroy` deletes the vault and drains only the budget *ledger*, not the address-balance *pool*; any undrained coin strands forever. It's **structurally unguardable on-chain** (Move can't enumerate runtime coin types) and heavily documented (DANGER block + drain ritual). Decision needed: consciously accept the limitation, or invest in a 2-step / SDK-enforced teardown before mainnet.
- [x] **Add a regression test for the strand** — pin the documented behavior: `destroy` with an undrained pool succeeds and the funds become unreachable.
- [x] **`set_allowance` takes an unvalidated bare `cap_id`** (informational) — a mistyped id silently provisions a different budget with no error signal on the create branch. Owner-only/self-inflicted and documented; optionally add an example showing the `AllowanceSet.was_created` confirmation pattern.

No action (noted): bearer-cap exposure and `delete_orphaned_cap` live-vault stranding are by-design and prominently documented. CI is green for `allowance`; the only red check is the pre-existing `Test math/core` timeout, unrelated to this PR.

**Comment by bidzyyys** [comment]
### Test coverage (`/sui-btt`, core `spend_vault`)

Coverage is **tight** — among the most thoroughly-tested modules in the repo (174 tests). All 8 error codes have `#[expected_failure]` paths; all 11 events are pinned with full-struct field assertions (not count-only); both sentinels (UNLIMITED budget incl. "never decrements", no-expiry) are covered; expiry is pinned at the closed boundary (exact-ms abort + one-ms-before success); and cross-coin/cap/vault isolation, lifecycle (9 destroy tests), reads, fund, revoke (incl. live-cap strand) and composability each have dedicated coverage. No `❌` on any invariant-backed leaf.

Two candidate gaps:

- [x] **`destroy` with an undrained pool isn't pinned** (Medium) — the documented DANGER (destroy strands pool funds) has no regression test. Caveat: after `destroy` the UID is gone, so `test_scenario` likely can't observe the stranded address balance; the achievable assertion is "destroy with a funded pool succeeds" (pins that no on-chain guard blocks it). If the harness can't see the post-destroy balance, this is fairly marked Deferred.
- [x] **`withdraw_all` on an empty settled pool** (Low) — the `amount == 0` branch (returns `balance::zero`, no `redeem_funds`, no abort) doesn't appear to have a dedicated test; verify it isn't covered under another name before adding.

Not a gap (by design): `InsufficientFundsForWithdraw` is an execution status, not a Move abort code, so it's intentionally not an `#[expected_failure(abort_code=...)]` leaf (matchable only via effects/dry-run).

**Comment by kosedogus** [comment]
> **destroy permanently strands an undrained pool**

Limitation is accepted and documented.

> **withdraw_all on an empty settled pool** 

Not unit-testable because we can't obtain `AccumulatorRoot` which is needed for `withdraw_all` function.

Others are resolved with [49eb62f](https://github.com/OpenZeppelin/contracts-sui/pull/402/changes/49eb62f92f027cbe99b206bd4960121f20e0cf69)

File: contracts/allowance/sources/spend_vault.move [hunk 4]
```diff
@@ -0,0 +1,1539 @@
...
+/// end one coin of a grant (owner side)          revoke<T> (idempotent; returns was_present)
+/// end an entire cap (owner side)                revoke_all (whole-cap kill)
+/// end a grant (spender side)                    renounce (consumes cap, all coins)
+/// dispose an orphaned cap (vault destroyed)     delete_orphaned_cap (prefer renounce if vault still live)
+/// recover a stray Coin sent to the vault        squash<T> (permissionless)
+/// emergency stop                                revoke_all (tx1), then withdraw_all<T> (tx2, retry-safe)
+/// tear down the vault                           withdraw_all<T> every coin, THEN destroy
+/// ```
+///
+/// #### Core semantics
```
Comment (ericnordelo) [inline contracts/allowance/sources/spend_vault.move:46]: These reads too verbose and complex for human. It feels like it can be simplified / improved for readability. Applies to the rest of the PR too, and something we should put in our skills too cc @bidzyyys

**Review by ericnordelo — COMMENTED** [review]
Looking great in general @kosedogus. Left just a general comment about doc readability, and mentioning here that we probably should follow up with the Sui team on being able to unit test AcumulatorRoot for address balances.

File: contracts/allowance/sources/spend_vault.move (hunk 4 above)
Comment (bidzyyys) [reply contracts/allowance/sources/spend_vault.move:46]: @ericnordelo / @kosedogus feel free to add this in this PR ✅

File: contracts/allowance/sources/spend_vault.move [hunk 5]
```diff
@@ -0,0 +1,1539 @@
...
+
+// === Structs ===
+
+/// Shared, UNTYPED escrow + per-`(cap, coin)` allowance ledger. One vault holds
+/// N coin types at once.
+///
+/// `key`-only by design: a Vault returned by `new` cannot be silently
+/// discarded (no `drop`), and external modules cannot wrap or re-share it (no
+/// `store`). Its lifecycle is exactly `new -> share` or `new -> destroy`,
+/// controlled solely by this module.
```
Comment (0xNeshi) [inline contracts/allowance/sources/spend_vault.move]: Seems redundant, it's just describing `key` properties.

File: contracts/allowance/sources/spend_vault.move [hunk 6]
```diff
@@ -0,0 +1,1539 @@
...
+///
+/// The pool is NOT a field here: per-coin funds live as
+/// object-owned address balances at `object::id(v).to_address()`.
```
Comment (0xNeshi) [inline contracts/allowance/sources/spend_vault.move]:
```suggestion
/// object-owned address balances at `object::id_address(&v)`.
```
nit: https://docs.sui.io/references/framework/sui_sui/object#sui_object_id_address

File: contracts/allowance/sources/spend_vault.move [hunk 7]
```diff
@@ -0,0 +1,1539 @@
...
+/// Key-only therefore does not conserve the escrow directly: it protects
+/// `id` (the `&mut v.id` spend authority) and the ledger, and forces
+/// every teardown through `destroy` (the one path where the drain discipline
+/// can be required).
+///
+/// - `allowances`: a `LinkedTable` (not `Table`) so `destroy`/`revoke_all`/
```
Comment (0xNeshi) [inline contracts/allowance/sources/spend_vault.move]:
```suggestion
/// - `allowances`: a `LinkedTable` so `destroy`/`revoke_all`/
```
redundant

File: contracts/allowance/sources/spend_vault.move [hunk 8]
```diff
@@ -0,0 +1,1539 @@
...
+    granted_coin_types: VecSet<TypeName>,
+}
+
+/// Composite ledger key: one entry per `(cap, coin type)`. Must be
+/// `copy + drop + store` to be a `LinkedTable` key: `store` to live in the
+/// table, `copy` to rebuild the same key for lookup without consuming a held
+/// value, `drop` to discard a lookup key that is not inserted. Well-formed
+/// because both components (`ID`, `TypeName`) are themselves copy+drop+store.
+/// `coin_type` is always `type_name::with_defining_ids<T>()`.
+public struct BudgetKey has copy, drop, store {
```
Comment (0xNeshi) [inline contracts/allowance/sources/spend_vault.move]:
```suggestion
/// Composite ledger key: one entry per `(cap, coin type)`.
/// `coin_type` is always `type_name::with_defining_ids<T>()`.
public struct BudgetKey has copy, drop, store {
```
Redundant, describe the properties of attributes.

File: contracts/allowance/sources/spend_vault.move [hunk 9]
```diff
@@ -0,0 +1,1539 @@
...
+/// single source of truth for the grant's state (the cap carries no budget).
+/// The coin type is in the `BudgetKey`, not here, so the entry holds
+/// exactly two scalars and one cap has N independent `Allowance` values.
+///
+/// `remaining`: `u64::MAX` is the UNLIMITED sentinel (never decremented);
+/// `0` is a live-but-suspended entry; anything else is the raw drawable
+/// budget. `expires_at_ms`: `u64::MAX` is the NO-EXPIRY sentinel; any finite
+/// value must be strictly future at `set_allowance` time. `store` lets it live
+/// as a `LinkedTable` value; `drop` allows clean disposal during `pop_front`
+/// drains.
```
Comment (0xNeshi) [inline contracts/allowance/sources/spend_vault.move]: These notes should be placed as field-level docs above relevant fields

File: contracts/allowance/sources/spend_vault.move [hunk 10]
```diff
@@ -0,0 +1,1539 @@
...
+/// #### Parameters
+/// - `ctx`: Transaction context, used to allocate the Vault and OwnerCap UIDs
+///   and to record the creator.
+///
+/// #### Returns
+/// - The new `Vault` (consume it with `share` or `destroy`) and its sole
+///   `OwnerCap`, both by value.
+///
+/// #### Aborts
+/// Never. Emits `VaultCreated { vault_id, owner_cap_id, creator }`.
```
Comment (0xNeshi) [inline contracts/allowance/sources/spend_vault.move]:
```suggestion
///   `OwnerCap`, both by value.
```
No need to document when no aborts can occur. Remove similar occurrances in other functions.

File: contracts/allowance/sources/spend_vault.move [hunk 11]
```diff
@@ -0,0 +1,1539 @@
...
+public fun new(ctx: &mut TxContext): (Vault, OwnerCap) {
+    let vault_uid = object::new(ctx);
+    let vault_id = vault_uid.to_inner();
+
+    let vault = Vault {
+        id: vault_uid,
+        allowances: linked_table::new<BudgetKey, Allowance>(ctx),
+        granted_coin_types: vec_set::empty<TypeName>(),
+    };
```
Comment (0xNeshi) [inline contracts/allowance/sources/spend_vault.move]:
```suggestion
    let vault = Vault {
        id: object::new(ctx),
        allowances: linked_table::new<BudgetKey, Allowance>(ctx),
        granted_coin_types: vec_set::empty<TypeName>(),
    };
    
    let vault_id = object::id(&vault);
```
nit: more idiomatic

File: contracts/allowance/sources/spend_vault.move [hunk 12]
```diff
@@ -0,0 +1,1539 @@
...
+        vault_id,
+        owner_cap_id: object::id(&owner_cap),
+        creator: ctx.sender(),
+    });
+
+    (vault, owner_cap)
+}
+
+/// Share the Vault. Module-only entry point: `Vault` omits `store`, so external
+/// modules cannot share it another way.
```
Comment (0xNeshi) [inline contracts/allowance/sources/spend_vault.move]:
```suggestion
/// Share the Vault.
```
nit: seems redundant, again stating ability properties

File: contracts/allowance/sources/spend_vault.move [hunk 13]
```diff
@@ -0,0 +1,1539 @@
...
+///
+/// Must run in the same tx as `new`; there is no deferred-share path. After
+/// `share`, the Vault is addressable as a shared input only in subsequent
+/// transactions, so all same-PTB fund / grant / embed steps must precede it. No
+/// event: sharing is platform-visible.
+///
+/// #### Aborts
+/// Never.
```
Comment (0xNeshi) [inline contracts/allowance/sources/spend_vault.move]:
```suggestion
```

File: contracts/allowance/sources/spend_vault.move [hunk 14]
```diff
@@ -0,0 +1,1539 @@
...
+/// EVERY ledger entry (recovering each per-entry storage rebate), deletes
+/// both UIDs, and returns NOTHING (it cannot return N heterogeneous per-coin
+/// balances). The drain is O(n) in live entries; for a very large ledger,
+/// batch-`revoke` first to spread gas across txs. Teardown is never blockable
+/// by spender state. `VaultDestroyed` is the terminal event for every entry
+/// under this vault_id.
+///
+/// #### Parameters
+/// - `v`: The vault to tear down (consumed by value).
+/// - `cap`: The OwnerCap bound to `v` (consumed by value).
```
Comment (0xNeshi) [inline contracts/allowance/sources/spend_vault.move]:
```suggestion
/// - `v`: The vault to tear down.
/// - `cap`: The OwnerCap bound to `v`.
```

File: contracts/allowance/sources/spend_vault.move [hunk 15]
```diff
@@ -0,0 +1,1539 @@
...
+/// - `ctx`: Transaction context, used to attribute the `VaultDestroyed` event.
```
Comment (0xNeshi) [inline contracts/allowance/sources/spend_vault.move]:
```suggestion
/// - `ctx`: Transaction context.
```
no need to state _why it's needed_, just describe what it _is_

File: contracts/allowance/sources/spend_vault.move [hunk 16]
```diff
@@ -0,0 +1,1539 @@
...
+///
+/// #### Aborts (in order)
```
Comment (0xNeshi) [inline contracts/allowance/sources/spend_vault.move]:
```suggestion
/// #### Aborts
```
remove such parantheses

File: contracts/allowance/sources/spend_vault.move [hunk 17]
```diff
@@ -0,0 +1,1539 @@
...
+/// 1. `EWrongOwnerCap`: cap bound to a different Vault. Only abort.
```
Comment (0xNeshi) [inline contracts/allowance/sources/spend_vault.move]:
```suggestion
/// - `EWrongOwnerCap` if cap is bound to a different Vault.
```
1. convention is to use dashes
2. conventional format

Apply this to other functions as well

File: contracts/allowance/sources/spend_vault.move [hunk 18]
```diff
@@ -0,0 +1,1539 @@
...
+
+    c.into_balance().send_funds(object::id(v).to_address());
+
+    event::emit(Deposited {
+        vault_id: object::id(v),
+        coin_type: type_name::with_defining_ids<T>(),
+        amount,
+        depositor: ctx.sender(),
+    });
+}
```
Comment (0xNeshi) [inline contracts/allowance/sources/spend_vault.move:553]:
```suggestion
public fun deposit<T>(v: &Vault, c: Coin<T>, ctx: &mut TxContext) {
    v.deposit_balance(c.into_balance());
}
```

File: contracts/allowance/sources/spend_vault.move [hunk 19]
```diff
@@ -0,0 +1,1539 @@
...
+/// - `b`: The `Balance<T>` to deposit (consumed by value).
+/// - `ctx`: Transaction context, used to attribute the `Deposited` event.
+///
+/// #### Aborts (in order)
+/// 1. `EZeroAmount`: `b.value() == 0`.
+public fun deposit_balance<T>(v: &Vault, b: Balance<T>, ctx: &mut TxContext) {
+    let amount = b.value();
+    assert!(amount > 0, EZeroAmount);
+
+    b.send_funds(object::id(v).to_address());
```
Comment (0xNeshi) [inline contracts/allowance/sources/spend_vault.move]:
```suggestion
    b.send_funds(object::id_address(v));
```

File: contracts/allowance/sources/spend_vault.move [hunk 20]
```diff
@@ -0,0 +1,1539 @@
...
+    event::emit(SpenderCapMinted {
+        vault_id: object::id(v),
+        cap_id,
+        by: ctx.sender(),
+    });
+
+    spender_cap
+}
+
+/// Owner-only. UPSERT the `(cap_id, T)` budget: create it if absent, else
```
Comment (0xNeshi) [inline contracts/allowance/sources/spend_vault.move]:
```suggestion
/// UPSERT the `(cap_id, T)` budget: create it if absent, else
```
nit: obvious from function signature

File: contracts/allowance/sources/spend_vault.move [hunk 21]
```diff
@@ -0,0 +1,1539 @@
...
+/// - `expected`: Optional CAS guard; `Some(e)` proceeds only if the entry exists
+///   and its current `remaining == e`, `None` is unconditional.
+/// - `clock`: Reference to the Sui `Clock`, used to validate `new_expires_at_ms`.
+/// - `ctx`: Transaction context, used to attribute the `AllowanceSet` event.
+///
+/// #### Aborts (in order)
+/// 1. `EWrongOwnerCap`: cap bound to a different Vault.
+/// 2. `EExpiryInPast`: finite `new_expires_at_ms <= now`.
+/// 3. `EUnexpectedAllowance`: CAS provided and the entry is absent or its
+///    current `remaining` differs.
```
Comment (0xNeshi) [inline contracts/allowance/sources/spend_vault.move]: Sanity check: run `/code-quality` skill

File: contracts/allowance/sources/spend_vault.move [hunk 22]
```diff
@@ -0,0 +1,1539 @@
...
+/// #### Parameters
+/// - `v`: The vault whose pool receives the recovered stray.
+/// - `c`: The `Receiving<Coin<T>>` ticket for the stray coin sent to `v`.
+/// - `ctx`: Transaction context, used to attribute the `Squashed` event.
+///
+/// #### Aborts
+/// Never on pool/ledger state. (The framework `public_receive` can abort on an
+/// invalid or stale `Receiving` ticket: a framework guarantee, not a library
+/// abort.)
+public fun squash<T>(v: &mut Vault, c: Receiving<Coin<T>>, ctx: &mut TxContext) {
```
Comment (0xNeshi) [inline contracts/allowance/sources/spend_vault.move:1110]: 1. Why the name `squash`? Wouldn't `receive_and_deposit` be clearer?
2. Consider accepting a vector of coins, i.e. `vector<Receiving<Coin<T>>>`, as it would enable depositing multiple/all `public_transfer`'d coins at once.

File: contracts/allowance/sources/spend_vault.move [hunk 23]
```diff
@@ -0,0 +1,1539 @@
...
+    let coin = transfer::public_receive(&mut v.id, c);
+    let amount = coin.value();
+    coin.into_balance().send_funds(vault_id.to_address());
+
+    event::emit(Squashed {
+        vault_id,
+        coin_type: type_name::with_defining_ids<T>(),
+        amount,
+        by: ctx.sender(),
+    });
```
Comment (0xNeshi) [inline contracts/allowance/sources/spend_vault.move:1122]: Do we need the `Squashed` event? I kind of see the point and I don't see it at the same time. How useful would this event really be?

File: contracts/allowance/sources/spend_vault.move [hunk 24]
```diff
@@ -0,0 +1,1539 @@
...
+///    balance is below `amount`. A funds-accumulator execution status raised at
+///    `redeem_funds` (surfaced in effects / dry run / SDK), NOT a Move `#[error]`
+///    code you can match with `expected_failure(abort_code = ...)`, and not one
+///    of this module's codes.
+public fun withdraw<T>(
+    v: &mut Vault,
+    cap: &OwnerCap,
+    amount: u64,
+    ctx: &mut TxContext,
+): Balance<T> {
```
Comment (0xNeshi) [inline contracts/allowance/sources/spend_vault.move:1162]: Is it idiomatic in Move to return `Balance<T>` or `Coin<T>`? Didn't research deep enough, so curious

**Review by 0xNeshi — APPROVED** [review]

File: contracts/allowance/sources/spend_vault.move (hunk 22 above)
Comment (kosedogus) [reply contracts/allowance/sources/spend_vault.move:1110]: This is not the expected usecase for the module. We don't want users to send Coin directly to vault. It is a safety just in case someone mistakenly sends to rescue the value. Name also came from Manos' suggestion for this safety function. We can change the name if it doesn't seem good, but `receive_and_deposit` would convey an idea that this is an intended path, which isn't, so I wouldn't suggest that.

File: contracts/allowance/sources/spend_vault.move (hunk 24 above)
Comment (kosedogus) [reply contracts/allowance/sources/spend_vault.move:1162]: Module became fully Address Balance native with the recent changes. Mysten Labs specifically asked for not returning Coin and returning Balance instead.

File: contracts/allowance/sources/spend_vault.move (hunk 23 above)
Comment (kosedogus) [reply contracts/allowance/sources/spend_vault.move:1122]: Not 100% necessary indeed considering we can't fully follow how much is deposited to any vault since there is also option to send via `send_funds`. Though I would keep it for consistency. Can't emit `Deposited` here considering it has a `depositor` field and event would misrepresent recoverer as a depositor.

**Comment by kosedogus** [comment]
Update on the accumulator coverage. `create_for_testing` for `AccumulatorRoot` was added to Sui main, but it isn't in any released version yet. It's on the `sui-v1.74.0-release` branch, so it comes with 1.74, not a 1.73 patch. Once it is released and we bump `SUI_VERSION` to 1.74, I can add unit tests for `balance_value`, `spendable_now`, and `withdraw_all`'s empty-pool + owner-gate branches, which closes most of the gap.

The one branch that stays is `withdraw_all`'s funded drain. It reads `settled_funds_value`, and that value is only written by the system settlement step, which test_scenario doesn't run, so in a unit test you can only hit the zero case. Sui itself tests settled reads through the transactional adapter, not move unit tests. I will raise a request from ML for a test-only setter so we can eventually pull that branch into unit tests too.

**Review by ericnordelo — APPROVED** [review]
